### DISCUSSION TITLE
Future Ideas and possible features (2024)

### DISCUSSION THREAD

**Discussion by ghiscoding** [ghiscoding/slickgrid-universal#1527] [Ideas]
#### Below is a list of ideas, there's a good change that it will never come to fruition but I'm just listing them as possible ideas anyway

- [x] 1. Vue port (PR #1753)
- [x] 2. row span, see [comments](https://github.com/6pac/SlickGrid/issues/381#issuecomment-951094855) and GerHobbelt initial [commit](https://github.com/GerHobbelt/SlickGrid/commit/397eab80ed43a7048b8c6554ffb03a4d7b74a2e8) (ref PR #1798)
  - SyncFusion has a nice [example](https://ej2.syncfusion.com/demos/#/bootstrap5/grid/row-spanning.html)
- [x] 3. we only have 1 preheader, so this mean we can't use simple features that use the same preheader because of the conflicts (DraggableGrouping/HeaderGrouping)
   - ref PR #1556
- [x] 4. ~possible use of Web Worker for Export Services to avoid blocking main thread while downloading~
   - gave it a try but it doesn't look feasible since Web Workers don't even work with binding scope, it only works with local code that has no connection (scope) to the outside which seems impossible in here since we need to use the SlickGrid/DataView variable bindings to read every row and all kind of other things
- [x] 5. Infinite Scroll for Backend Services (see Ag-Grid [Infinite Row Model](https://www.ag-grid.com/angular-data-grid/infinite-scrolling/)) 
  - PR #1609
  - PR #1628 
- [x] 6. ~try to use SortableJS ESM import and see if it helps with build size~
   - `import Sortable from 'sortablejs/modular/sortable.core.esm.js';` (see SortableJS [Getting Started](https://github.com/SortableJS/Sortable?tab=readme-ov-file#getting-started))
   - make sure Column Reordering and Draggable Grouping also still works
   - currently makes 2/3 of unit tests to fail in PR #1621 because Jest only supports CJS, so I'll cancel this for now but might revisit if migrating to Vitest with full ESM
   - **EDIT**: I gave it another try after migrating to Vitest and `sortable.core.esm.js` doesn't play well with current Types `@types/sortablejs` because the names don't match. Considering that this is a pain to get proper Types, I think I'll just keep current approach and 1kb bigger is not that much of a difference. The hope is that one day, SortableJS offers better ESM support (but I won't put much hope on this happening anytime soon sadly)
 - [x] 7. drag rows ~between 2 grids~ and/or drag row to outside container like a recycle bin
    - [x] 1. ~[between 2 grids](https://6pac.github.io/SlickGrid/examples/example-drag-row-between-grids.html)~ (won't do since this requires a new [plugin](https://github.com/6pac/SlickGrid/blob/master/src/plugins/slick.crossgridrowmovemanager.ts) and no one ever requested this feature), alternatively any user could simply create their own plugin themselves. 
    - [x] 2. [drag rows to recycle bin](https://6pac.github.io/SlickGrid/examples/example9-row-reordering.html)
- [x] 8. ~stop using Jest Global types, see this potential [PR](https://github.com/FloEdelmann/vue-ts-types/pull/532/files) of another project to handle it~ Migrate to Vitest instead
   - ref PR #1663
- [x] 9. can we further improve Date Sorting speed (perf)? (see if we can use other sorting type for large dataset, i.e. quick sort)
  - ref PR #1691
  - see these articles [Sorting Algorithms in JavaScript or TypeScript](https://dev.to/bugudiramu/a-developers-guide-to-sorting-algorithms-2kl9) / [Introduction to Sorting Algorithms in JavaScript](https://dev.to/alexmercedcoder/introduction-to-sorting-algorithms-in-javascript-b60)
- [x] 10. custom pagination is not currently replaceable by the user, but it probably should be eventually to be fully customizable 
- [x] 11. add `silenceWarnings` to all ports in next version only
- [x] 12. I think that `GroupingAndColspanService` should be renamed to `HeaderGroupingService` since there's nothing related to colspan (the colspan was actually meant to represent that header can span across multiple columns, but SlickGrid always called it Header Grouping, so let's use that name instead in the next `minor` version)
- [x] 13. we should probably add peer deps to all ports, for example: `"peerDependencies": { "react": ">= 18.0.0" }`
- [x] 14. currently Grid Option are required but I think it should be made optional
- [x] 15. add new example to showcase data CSV import to build a grid (PR #1772)
- [x] 16. can we replace `fetch-jsonp` with native Fetch? Maybe this [SO](https://stackoverflow.com/a/41149634/1212166)
- [x] 17. auto-resize height by dataset length instead of full window height available. From grid row height & dataset length we can calculate grid height

**Comment by ghiscoding** [comment]
> 27. in Angular-Slickgrid, we should probably:
> * [ ]  migrate from Jest to Vitest using Analog (see [Faster testing with Angular and Vitest](https://dev.to/brandontroberts/faster-testing-with-angular-and-vitest-274n))

@zewa666 I wouldn't mind the help on this if you want to take a look at it in Angular-Slickgrid, it will be required before moving the project in here (I just don't want to have to test frameworks in here)

**Reply by zewa666** [reply]
we've got our apps converted to use both vite and vitest. so shoot if I can help

**Reply by ghiscoding** [reply]
if you can provide a PR in Angular-Slickgrid to migrate to Vitest, I'd be more than happy 😉

**Reply by zewa666** [reply]
cant promise anything about when as I'm currently burried with work, but I will asap

**Reply by ghiscoding** [reply]
np, it's not that much of a rush since I'm not planning major release until around April (near NodeJS release) so..
